Improve watcher

nyarly opened the pull request:
Expands on #113, fixes the Vim test by debouncing notifications internally (which is easier because we only care about touched paths.)

Profpatsch committed:
fix(build.ninja): make Cargo.nix indirect input

that way it’s not passed as $in to the build rule, which broke the nix
invocation.

Profpatsch committed:
chore(Cargo.toml): Update crossbeam_channel and notify

crossbeam_channel should work as before, there were no functional
changes.

inotify is a major version, so it changes watching behaviour by a bit,
making some of the tests fail.

Profpatsch commented on src/watch.rs:
```diff
@@ -51,14 +53,41 @@ impl WatchPathBuf {
+        let (watch_tx, watch_rx) = chan::unbounded();
+
+        let filter = Filter {
+            notify: Watcher::new(ntfy_tx, Config::default())?,
```
Is there a reason why the new watcher uses a `30s` poll interval, as opposed to the `0.1s` from before?

Profpatsch replied:
I think we can probably increase it up to 3 seconds to take some pressure off of bigger projects, but 30s would kind of destroy the idea imo.

nyarly replied:
The duration isn't important. I don't have a problem doing something other than the default config, but:

* We aren't using Notify's debounce anymore, so that delay isn't in the config
* I think that everywhere we care about, Notify is using OS level FS events, not polling. Test behavior bears that out.

Profpatsch replied:
yeah, true, I figured out it’s only relevant for the poll backend.

Profpatsch commented on src/lib.rs:
```diff
@@ -216,21 +216,4 @@ mod tests {
         )
     }
-
-    #[test]
-    fn test_locate_config_file() {
-        let mut path = PathBuf::from("shell.nix");
-        let result = AbsPathBuf::new_from_current_directory(&path);
-        assert_eq!(
-            result
-                .unwrap()
-                .expect("Should find the shell.nix in this projects' root"),
-            AbsPathBuf::new(PathBuf::from(env!("CARGO_MANIFEST_DIR")))
-                .unwrap()
-                .join("shell.nix")
-        );
-        path.pop();
-        path.push("this-lorri-specific-file-probably-does-not-exist");
-        assert_eq!(None, AbsPathBuf::new_from_current_directory(&path).unwrap());
-    }
```
why remove this test?

nyarly replied:
As a result of rebasing, it was duplicated, so this just removes a copy of the test.

Profpatsch commented on src/watch.rs:
```diff
@@ -51,14 +53,41 @@ impl WatchPathBuf {
+            logger: logger.clone(),
+        };
+
+        thread::spawn(move || filter.run());
```
We shouldn’t just spawn threads and ignore their thread handles

nyarly replied:
That was my first instinct but we don't have anything to wait on the thread for (all its results are streamed out), so there's little value to `join`ing.

Once the Watcher gets dropped, its side of the channels are dropped, and the Filter will get those errors (e.g. RecvError) and exit its loop.

In the end, I couldn't find a sensible way to make the captured thread handle not dead code, which is why I ignored it.

Profpatsch replied:
Check out the code, I added `Async::run_with_stop_signal` to kill the event loop on drop.

Profpatsch committed:
feat(Async): add Async::run_with_stop_signal

This function allows us to spawn a cooperative async, which is asked
to stop when the async is dropped.

We don’t add a timeout when waiting in the drop, meaning if the inner
async ignores the channel it might block execution.

Profpatsch commented on src/watch.rs:
```diff
@@ -71,33 +100,75 @@ struct FilteredOut<'a> {
+
+    fn buffer_notify(&self, mut touched: Vec<PathBuf>) -> Vec<PathBuf> {
+        let until = Instant::now() + Duration::from_millis(200);
+        loop {
+            match self.notify_rx.recv_deadline(until) {
+                Ok(event) => {
+                    if let Some(mut paths) = self.process_watch_events(event) {
+                        touched.append(&mut paths);
+                    }
+                }
+                Err(_) => {
+                    touched.dedup();
+                    return touched;
+                }
+            }
```
I don’t think it’s a good idea to have a loop-in-a-loop that only watches for one message. The outer message `select!` will not be live while this blocks.

Profpatsch replied:
We already listen for filesystem_events_rx outside of this helper, so blocking *only* on that means we don’t get any messages for e.g. new filepaths to watch, which might run over the channel bounds of those channels.

nyarly replied:
We're watching on a deadline here, so this is where the 100ms goes to live in this code. Effectively, we're debouncing now at the watched path level, rather than on an Event level, which is what was required to solve the "vim-wait" test (although I think that covers a more general use case, of any complex file manipulation.)

Indeed, watching new paths would be the only messages that have to wait while we debounce, and even in the perverse case where many new paths were registered in 100ms, the channel is unbounded. 

There's a possible edgecase where a single path is reported as needing to be watched many times, so if we were consuming off the watch channel continuously, we could effectively deduplicate them as they were added to the HashMap, and in the meantime they consume memory in the channel. It seems like a particularly low risk issue though.

Profpatsch replied:
I think the main issue I had with this is that, instead of consuming events as fast as they come in, we’d now add a fixed 200ms delay to our event loop. So if the user sends a new path to be watched, we’d not react for 200ms while in the inner loop.

Profpatsch replied:
But check out the new code, it does not have that problem.

Profpatsch committed:
feat(watcher): Use notify_debouncer_full

Based on code by @nyarly, this debounces watch events before filtering
file paths.

`notify_debouncer_full` will do the hard part of making sure spurious
events are sufficiently removed before we match on them. From the
docs:

* Only emits a single Rename event if the rename From and To events can be matched
* Merges multiple Rename events
* Takes Rename events into account and updates paths for events that occurred before the rename event, but which haven't been emitted, yet
* Optionally keeps track of the file system IDs all files and stiches rename events together (FSevents, Windows)
* Emits only one Remove event when deleting a directory (inotify)
* Doesn't emit duplicate create events
* Doesn't emit Modify events after a Create event

Profpatsch commented:
The upstream `notify` library has a sister library called https://docs.rs/notify-debouncer-full/latest/notify_debouncer_full/ in its newer versions, so I rewrote some of the filter code to use that instead, and it seems to fix the vim test without any need for manual debouncing on our part! PTAL.

Profpatsch committed:
fix(watcher): rephrase macos event drop test fix

`macos_eat_late_notifications` is now less straightforward to
implement, because we separated `Watch` from `Filter` and don’t expose
the filter channel anymore.

So instead let’s pass an optional duration, during which the first
event is dropped.

This somewhat complicates the event loop, but it can’t be helped, I
think??

Profpatsch commented:
Also lol, searching for that macos panic only brings up https://users.rust-lang.org/t/mysterious-panic-in-macos-cargo-test/77997/1

Profpatsch committed:
feat(tests): add test name to logger messages

We put every test in a sublogger and add its test name, to aid in
debugging.

Profpatsch committed:
feat(watcher): add test name to test dirs

Profpatsch committed:
feat(watcher): log subparts of rename_over_vim

Profpatsch committed:
fix(watcher): only watch /bar to work around macos notify race

MacOS has some delay between any change and when the watcher actually
notices it. So it would always show the creation of foo when we want
to make sure `bar` does not get watched, leading to the test failing.

Instead, filter out every file that is not `/bar` here, which should
be fine.

Profpatsch commented:
Actually I *think* the rename_over_vim test is broken on *Linux* not mac; we are watching the parent dir, so I’d expect it to notify about the creation of `bar`! Which is the part that fails on Mac.

Profpatsch commented:
Ah, nvm, it’s not that, it should be filtered out by the path filtering, so not generate an interesting event, which is the correct case. It’s hard to know with the panic unwinding being broken on current MacOS runners … probably relevant to that is https://github.com/rust-lang/rust/issues/113783

I’d just disable the test on macos for now and hope it fixes itself … maybe
